viewer/src/store.ts

```diff
@@ -1,26 +1,71 @@
 import Engine, {
   AdvTechTilePos,
   Booster,
+  Command,
   Federation,
   GaiaHex,
+  LogEntry,
   Player,
+  PlayerEnum,
   ResearchField,
   TechTilePos,
 } from "@gaia-project/engine";
 import { CubeCoordinates } from "hexagrid";
+import { findLast, findLastIndex } from "lodash";
 import Vue from "vue";
 import Vuex from "vuex";
 import { GameContext } from "./data";
 
 Vue.use(Vuex);
 
+function ownTurn(move: string): boolean {
+  if (move == null) {
+    return false;
+  }
+  const outOfTurn = [Command.ChargePower, Command.BrainStone, Command.ChooseIncome, Command.Decline];
+  return !outOfTurn.some((command) => move.includes(command));
+}
+
+function movesToHexes(data: Engine, moves: string[]): GaiaHex[] {
+  return moves.flatMap((move) => {
+    const args = data.parseMove(move).args;
+    const number = args.indexOf("build");
+    if (number >= 0) {
+      const coord = args[number + 2].replace(".", "");
+      const hex = data.map.getS(coord);
+      return [hex];
+    }
+    return [];
+  });
+}
+
+function recentMoves(player: PlayerEnum, logEntries: LogEntry[], moveHistory: string[]) {
+  let last = logEntries.length;
+  let lastMove = moveHistory.length;
+  while (last > 0 && logEntries[last - 1].player === player) {
+    const move = logEntries[last - 1].move;
+    if (move) {
+      lastMove = move;
+    }
+    last--;
+  }
+
+  const firstMove = findLast(
+    logEntries.slice(0, last),
+    (logItem) => logItem.player === player && ownTurn(moveHistory[logItem.move])
+  )?.move;
+  return firstMove ? moveHistory.slice(firstMove + 1, lastMove) : [];
+}
```

Comment:
Move this to a separate file, something like "logic/recent-moves.ts". 

The test file can be "logic/recent-moves.spec.ts".